### ISSUE TITLE
Client disconnects after sending groupchat message

### ISSUE THREAD

**Issue by sschimmel** [xmppjs/xmpp.js#792]
Whenever I send a groupchat message the client closes the connection and reconnects. The groupchat message is send and received by an other client so sending the message seems to work.

I have setup a ejabberd server and created a client using xmpp.js. Connecting to the ejabberd server, sending presence, joinging MUC rooms and receiving the message history all works fine. 

When I send a groupchat message I see the message being sent, then the message being send back by the ejabberd server and then the xmpp.js client outputs a stream error. The error doesn't tell me why there's an error except for "bad format".

To test I have used conversejs.org's browser based client.  I logged in as a different user and joined the same MUC room as with my own client. I see the message being sent from my own client appear in the conversejs client so I know the message is send correctly. When I send a message from the conversejs client my own client also disconnects with the same error. So it seems there is a problem in receiving the groupchat message from the server. The weird thing is that there are no problems receiving the groupchat message history on joining a MUC room.

Below the console output from the different events:

```
⮊ <message type="groupchat" to="test4@conference.chat.dev.*****" id="3f606f8f-0308-4210-86ba-0b91962300f8" from="sebastiaan@chat.dev.****/1720985928900570160214546" xmlns="jabber:client"><body>Test</body></message> breadcrumbs.js:81
⮈ <message xmlns='jabber:client' xml:lang='en' to='sebastiaan@chat.dev****/1720985928900570160214546' from='test4@conference.chat.****/sebastiaan' type='groupchat' id='3f606f8f-0308-4210-86ba-0b91962300f8'><archived by='test4@conference.chat.****' id='1573481456001018' xmlns='urn:xmpp:mam:tmp'/><stanza-id by='test4@conference.chat.dev.****' id='1573481456001018' xmlns='urn:xmpp:sid:0'/><body>Test</body></message> breadcrumbs.js:81
⮊ <stream:error><bad-format xmlns="urn:ietf:params:xml:ns:xmpp-streams"/></stream:error> breadcrumbs.js:81
⮊ <close xmlns="urn:ietf:params:xml:ns:xmpp-framing"/> breadcrumbs.js:81
🛈 status closing breadcrumbs.js:81
⮈ <close xmlns='urn:ietf:params:xml:ns:xmpp-framing'/> breadcrumbs.js:81
🛈 status closing breadcrumbs.js:81
🛈 status disconnecting breadcrumbs.js:81
🛈 status disconnect breadcrumbs.js:81
```

I've spend several days at searching for a solution and I'm now stuck in solving this. It might be the client or it might be the ejabberd server generating incorrect messages. Before I install a different xmpp server I'd like to rule out that it's the client.

Any ideas / help are greatly appreciated :D

**Comment by sonnyp** [comment]
Might be repeating what you said but just in case, the third line indicates that it is xmpp.js sending a stream error.

Per specification, on stream error, both parties (server and client) close the connection.

So you should look into why `xmpp.js` is sending a stream error. There is only one place where this can happen: `_onData` https://github.com/xmppjs/xmpp.js/blob/master/packages/connection/index.js#L64

You could manually edit the file in your node_modules to log the input string and the error.

**Comment by sschimmel** [comment]
@sonnyp thanks for your help. I was able to fix the error with adding logging to _onData();

I'm typing this here while debugging :-) Helps me debug and it might help someone in the future. TL;DR; the error was in my own code and in the end catched in the _onData(). 

I added some console.logs() and it appears that the error is happening in my code that processes the message.

The logging added to the _onData() function.
```
async _onData(data) {
    console.log('XMPP, data received: ', data);
    const str = data.toString('utf8')
    this.emit('input', str)
    try {
      await this.parser.write(str)
      // eslint-disable-next-line no-unused-vars
    } catch (err) {
      console.log('XMPP, parser.write error: ', err);
      // https://xmpp.org/rfcs/rfc6120.html#streams-error-conditions-bad-format
      // "This error can be used instead of the more specific XML-related errors,
      // such as <bad-namespace-prefix/>, <invalid-xml/>, <not-well-formed/>, <restricted-xml/>,
      // and <unsupported-encoding/>. However, the more specific errors are RECOMMENDED."
      try {
        this._streamError('bad-format')
        // eslint-disable-next-line no-unused-vars
      } catch (err) {}
    }
  }
```

The message received by the _onData function that results in the stream error.
```
XMPP, data received:  <message xmlns='jabber:client' xml:lang='en' to='sebastiaan@chat.dev.****/1796433921420323802614818' from='test4@conference.chat.****/sebastiaan' type='groupchat' id='5bbb208e-3be2-4a49-9065-31696b6a5a40'><archived by='test4@conference.chat.****' id='1573548574154754' xmlns='urn:xmpp:mam:tmp'/><stanza-id by='test4@conference.chat.*****' id='1573548574154754' xmlns='urn:xmpp:sid:0'/><body>Test 2</body></message>
```

The error that is logged in the catch: 
```
XMPP, parser.write error:  TypeError: "stanza.getChild(...) is undefined"
    processGroupChatMessage index.js:133
    processStanza index.js:122
```

The error indicates that stanza.getChild is undefined. That error is thrown in processGroupChatMessage() which is called by processStanza(). Both are functions I have written.

The odd thing is that processGroupChatMessage is called by processStanza based on the content of the stanza that includes mulitple .getChild() statements.

processStanza()
```
processStanza(stanza) {
    console.info(stanza);

    if (stanza.is('message')) {
      if (stanza.getChild('body') && stanza.getChild('body').text()) {
        if (stanza.attrs.type === 'groupchat') {
          this.processGroupChatMessage(stanza);
        }
      }
    }
  }
```

The code for processGroupChatMessage():
```
processGroupChatMessage(stanza) {
    const chat = stanza.attrs.from.split('/')[0];
    const from = stanza.attrs.from.split('/')[1];
    const message = stanza.getChild('body').text();
    const msgId = stanza.getChild('stanza-id').attrs.id;
    const date = stanza.getChild('delay').attrs.stamp;

    const msg = {
      id: msgId,
      chat,
      from,
      message,
      date,
    };

    this.emit('NewGroupChatMessage', { message: msg });
  }
```

What was causing the error was that `delay` was missing from the message received and thus undefined. The error so it seems was ultimately catched in the _onData function.

**Comment by sonnyp** [comment]
@sschimmel thank you for sharing

I'm re-opening this because it demonstrates a serious bug in xmpp.js and I don't want to forget about it.

**Comment by sonnyp** [comment]
@sschimmel this was happening on React Native right?

**Comment by sonnyp** [comment]
Nevermind, can reproduce with Node.js

**Comment by sonnyp** [comment]
Simple test case;

```js
/* eslint-disable node/no-extraneous-require */

'use strict'

const {client, xml} = require('@xmpp/client')
const debug = require('@xmpp/debug')

const xmpp = client({
  service: 'ws://localhost:5280/ws',
  domain: 'localhost',
  resource: 'example',
  username: 'username',
  password: 'password',
})
debug(xmpp, true)
xmpp.reconnect.stop()

xmpp.on('stanza', stanza => {
  // Throws
  stanza.foo.hello()
})

xmpp.on('online', address => {
  xmpp.send(
    xml('message', {type: 'chat', to: address}, xml('body', {}, 'hello world'))
  )
})

xmpp.start().catch(console.error)
```

Outputs

```xml
IN
<iq xmlns="jabber:client" type="result" id="p5v4iu1sav">
  <bind xmlns="urn:ietf:params:xml:ns:xmpp-bind">
    <jid>
      username@localhost/example
    </jid>
  </bind>
</iq>
OUT
<stream:error>
  <bad-format xmlns="urn:ietf:params:xml:ns:xmpp-streams"/>
</stream:error>
status closing 
status online username@localhost/example
(node:34168) UnhandledPromiseRejectionWarning: Error: write EPIPE
    at afterWriteDispatched (internal/stream_base_commons.js:150:25)
    at writevGeneric (internal/stream_base_commons.js:133:3)
    at Socket._writeGeneric (net.js:768:11)
    at Socket._writev (net.js:777:8)
    at doWrite (_stream_writable.js:429:12)
    at clearBuffer (_stream_writable.js:528:5)
    at Socket.Writable.uncork (_stream_writable.js:325:7)
    at Sender.sendFrame (/home/sonny/simplo/xmpp/node_modules/ws/lib/sender.js:353:20)
    at Sender.send (/home/sonny/simplo/xmpp/node_modules/ws/lib/sender.js:272:12)
    at WebSocket.send (/home/sonny/simplo/xmpp/node_modules/ws/lib/websocket.js:354:18)
(node:34168) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:34168) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
status disconnect [object Object]
```

**Comment by sonnyp** [comment]
closing in favor of https://github.com/xmppjs/xmpp.js/pull/801

**Comment by sonnyp** [comment]
fixed in v0.9.1 https://github.com/xmppjs/xmpp.js/releases/tag/v0.9.1
